### PR TITLE
Fixed #30995 -- Allowed converter.to_url() to raise ValueError to indicate no match.

### PR DESCRIPTION
**Pull request by jcushman**
https://code.djangoproject.com/ticket/30995

### CODE REVIEW

File: django/urls/resolvers.py [hunk 1]
```diff
@@ -631,11 +631,15 @@ def _reverse_with_prefix(self, lookup_view, _prefix, *args, **kwargs):
...
-                        text_candidate_subs[k] = str(v)
+                try:
+                    for k, v in candidate_subs.items():
+                        if k in converters:
+                            text_candidate_subs[k] = converters[k].to_url(v)
+                        else:
+                            text_candidate_subs[k] = str(v)
+                except ValueError:
+                    # Allow converters to decline to match by raising ValueError.
+                    continue
```
Comment (jcushman) [inline django/urls/resolvers.py]: This try-except could be a little tighter, and just go around `text_candidate_subs[k] = converters[k].to_url(v)`, but then we need some extra moving parts to continue from the nested `for` loop.

File: tests/urlpatterns/polymorphic_urls.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    path('c/5/<uuid:value>/', views.empty_view, name='c'),
+
+    # ... re_path type
+    re_path(r'^d/1/([A-Z]+)/', views.empty_view, name='d'),
+    re_path(r'^d/2/([a-z]+)/', views.empty_view, name='d'),
+
+    # ... custom converter raising ValueError
+    path('e/1/<value>/', views.empty_view, name='e'),
+    path('e/2/<dynamic:value>/', views.empty_view, name='e'),
+]
```
Comment (jcushman) [inline tests/urlpatterns/polymorphic_urls.py]: All of these test cases already pass without the change to `resolvers.py`, except for the last one. I added the others to document the various ways Django currently handles polymorphic url name resolution.

Comment (felixxm) [reply tests/urlpatterns/polymorphic_urls.py]: Extra tests are fine but they should be moved to a separate commit.

File: docs/topics/http/urls.txt [hunk 3]
```diff
@@ -148,10 +148,17 @@ A converter is a class that includes the following:
   string into the type that should be passed to the view function. It should
   raise ``ValueError`` if it can't convert the given value. A ``ValueError`` is
   interpreted as no match and as a consequence a 404 response is sent to the
-  user.
+  user unless another matching path is able to convert the value.
```
Comment (felixxm) [inline docs/topics/http/urls.txt]: This clarification is unrelated, please move it to a separate commit.

File: django/urls/resolvers.py (hunk 1 above)
Comment (felixxm) [reply django/urls/resolvers.py]: Agreed, please move `try ... except` to the `text_candidate_subs[k] = converters[k].to_url(v)`.

File: tests/urlpatterns/polymorphic_urls.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
+from django.urls import path, re_path, register_converter
+
+from . import converters, views
+
+register_converter(converters.DynamicConverter, 'dynamic')
+
+
+# test reverse resolution of routes that share the same name, based on ...
```
Comment (felixxm) [inline tests/urlpatterns/polymorphic_urls.py]: I don't see much value in this comment please chop it.

File: tests/urlpatterns/polymorphic_urls.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+urlpatterns = [
+    # ... count
+    path('a/1/', views.empty_view, name='a'),
+    path('a/2/<value>/', views.empty_view, name='a'),
+
```
Comment (felixxm) [inline tests/urlpatterns/polymorphic_urls.py]: Please chop all these blank lines.

File: tests/urlpatterns/polymorphic_urls.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+from . import converters, views
+
+register_converter(converters.DynamicConverter, 'dynamic')
+
+
+# test reverse resolution of routes that share the same name, based on ...
+urlpatterns = [
+    # ... count
+    path('a/1/', views.empty_view, name='a'),
+    path('a/2/<value>/', views.empty_view, name='a'),
```
Comment (felixxm) [inline tests/urlpatterns/polymorphic_urls.py]: Can we use more specific names?, e.g.

- `a` -> `number_of_kwargs`,
- `b` -> `kwarg`,
- `c` -> `converter`,
- `d` -> `regex`, etc.

File: tests/urlpatterns/tests.py [hunk 7]
```diff
@@ -190,6 +190,58 @@ def test_nonmatching_urls(self):
                     resolve(url)
 
 
+@override_settings(ROOT_URLCONF='urlpatterns.polymorphic_urls')
+class PolymorphicNameTests(SimpleTestCase):
+    def test_resolve_polymorphic_names(self):
+        """
+        Given multiple routes that share the same name, reverse() should match to the expected route
+        based on argument count, keywords, and type.
```
Comment (felixxm) [inline tests/urlpatterns/tests.py]: Wrap docstrings at 79 chars.

File: docs/topics/http/urls.txt [hunk 8]
```diff
@@ -661,7 +668,11 @@ included at all).
 
 You may also use the same name for multiple URL patterns if they differ in
 their arguments. In addition to the URL name, :func:`~django.urls.reverse()`
-matches the number of arguments and the names of the keyword arguments.
+matches the number and type of arguments and the names of the keyword arguments.
+Specifically, :func:`~django.urls.reverse()` returns a URL for the last-defined
+path where the number of arguments or names of keyword arguments match; where
+the url converters do not raise ``ValueError``; and where the resulting URL
+matches the :func:`django.urls.path` or :func:`django.urls.re_path`'s regular expression.
```
Comment (felixxm) [inline docs/topics/http/urls.txt]: Wrap docs at 79 chars.

**Comment by jcushman** [comment]
Thanks, @felixxm! I believe all comments are resolved by the revised commits.

File: docs/topics/http/urls.txt [hunk 9]
```diff
@@ -148,10 +148,17 @@ A converter is a class that includes the following:
   string into the type that should be passed to the view function. It should
   raise ``ValueError`` if it can't convert the given value. A ``ValueError`` is
   interpreted as no match and as a consequence a 404 response is sent to the
-  user.
+  user unless another matching path is able to convert the value.
 
 * A ``to_url(self, value)`` method, which handles converting the Python type
-  into a string to be used in the URL.
+  into a string to be used in the URL. It should return a non-matching string
```
Comment (kezabelle) [inline docs/topics/http/urls.txt]: Is this documented elsewhere, or is it just an outcome of the way the implementation happens to work? If it's not documented elsewhere, is it sensible to make a _non-matching string_ a prescribed way of bailing out? By having that documented + the new option of `ValueError` raising, there'd then be **2** ways to do it.

(Counter point: someone returning a non-matching string in the future _could_ be surprised by the functionality, if it's not documented?)

Comment (jcushman) [reply docs/topics/http/urls.txt]: I think documenting both ways to do it makes sense? They're both useful in different situations.

**non-matching-string:** 

It's very convenient when it works, and it's the way that's used by all the built-in converters. For example, here's `UUIDConverter`:

```
class UUIDConverter:
    regex = '[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}'
    ...
    def to_url(self, value):
        return str(value)
```

If non-matching-string wasn't supported by an implementation, this would need an extra `if not re.match(self.regex+'$', str(value)): raise ValueError`, and so would every other converter -- it becomes redundant boilerplate for all converters. Seems reasonable to document that you don't need to do that. 

**raise ValueError**:

This one is more for clarity. It feels like abusing the type system to explicitly return a non-matching string when you mean "no match." For example, if we want to match instances of MyModel, this doesn't quite feel right:

```
class MyModelConverter(IntConverter):
    def to_url(self, value):
        if type(value) is not MyModel:
            return ''  # this isn't good practice ... right?
        return str(value.id)
```

The above solution hypothetically wouldn't work at all if there was no non-matching string -- not sure if this is a real situation though:

```
# could something like this ever be reasonable?

class MyModel(models.Model):
    title = models.CharField(length=255, blank=True)

class MyModelTitleConverter:
    regex = '.*'
    def to_url(self, value):
        if type(value) is not MyModel:
            raise ValueError
        return value.title

urls = [
    path('/<my_model:id>-<my_model_title:title>/`),
]
```

---

I don't know -- if we want to have Only One Way To Do It, I might document that a converter should "return an empty string (or other string that doesn't match the regular expression) to decline a match," since that seems like the indispensible one, and drop the special ValueError handling.

**Comment by felixxm** [comment]
@jcushman I pushed minor edits to the first two commits and moved them to a separate PR #12290.

File: docs/topics/http/urls.txt (hunk 9 above)
Comment (felixxm) [reply docs/topics/http/urls.txt]: IMO we shouldn't document returning a non-matching string (empty) as a supported way to decline a match. It's rather a side effect of the current implementation. I'm going to update docs changes.

File: docs/topics/http/urls.txt [hunk 10]
```diff
@@ -661,7 +668,12 @@ included at all).
 
 You may also use the same name for multiple URL patterns if they differ in
 their arguments. In addition to the URL name, :func:`~django.urls.reverse()`
-matches the number of arguments and the names of the keyword arguments.
+matches the number and type of arguments and the names of the keyword arguments.
```
Comment (felixxm) [inline docs/topics/http/urls.txt]: ` and type of argument ...` `reverse()` doesn't match type of arguments it is sth that path converters do. I will revert this change.

**Comment by felixxm** [comment]
@jcushman Thanks :+1: I pushed edits and rebased from master.

Welcome aboard :boat:

File: docs/releases/3.1.txt [hunk 11]
```diff
@@ -294,7 +294,8 @@ Tests
 URLs
 ~~~~
 
-* ...
+* :ref:`Path converters <registering-custom-path-converters>` can now raise
+  ``ValueError`` to indicate no match when reversing URLs.
```
Comment (carltongibson) [inline docs/releases/3.1.txt]: This should say "in ``to_url()``"

Comment (felixxm) [reply docs/releases/3.1.txt]: Added, thanks :+1:
